src/eventer/eventer_epoll_impl.c

```diff
@@ -287,31 +287,51 @@ static void eventer_epoll_impl_trigger(eventer_t e, int mask) {
       mtevL(eventer_deb, "Attempting to trigger already-registered event fd: %d cross thread.\n", fd);
     }
     /* mtevAssert(master_fds[fd].e == NULL); */
-    
+
     eventer_cross_thread_trigger(e,mask);
     return;
   }
   if(master_fds[fd].e == NULL) {
     lockstate = acquire_master_fd(fd);
     if (lockstate == EV_ALREADY_OWNED) {
       /* The incoming triggered event is already owned by this thread.
-	 This means our floated event completed before the current
-	 event handler even exited.  So it retriggered recursively
-	 from inside the event handler.
-
-	 Treat this special case the same as a cross thread trigger
-	 and just queue this event to be picked up on the next loop
-      */
+       * This means our floated event completed before the current
+       * event handler even exited.  So it retriggered recursively
+       * from inside the event handler.
+       *
+       * Treat this special case the same as a cross thread trigger
+       * and just queue this event to be picked up on the next loop
+       */
       eventer_cross_thread_trigger(e, mask);
       return;
     }
-    release_master_fd(fd, lockstate);
+    /*
+     * If we are readding the event to the master list here, also do the needful
+     * with the epoll_ctl.
+     *
+     * This can happen in cases where some event was floated and the float
+     * completed so fast that we finished the job in the same thread 
+     * that it started in.  Since we `eventer_remove_fd` before we float
+     * the re-add here should replace the fd in the epoll_ctl.
+     */
     master_fds[fd].e = e;
     e->mask = 0;
-    needs_add = 1;
+    struct epoll_event _ev;
+    memset(&_ev, 0, sizeof(_ev));
+    _ev.data.fd = fd;
+    spec = eventer_get_spec_for_event(e);
+    if(mask & EVENTER_READ) _ev.events |= (EPOLLIN|EPOLLPRI);
+    if(mask & EVENTER_WRITE) _ev.events |= (EPOLLOUT);
+    if(mask & EVENTER_EXCEPTION) _ev.events |= (EPOLLERR|EPOLLHUP);
+
+    mtevL(eventer_deb, "epoll_ctl(%d, add, %d)\n", spec->epoll_fd, fd);
+    if (epoll_ctl(spec->epoll_fd, EPOLL_CTL_ADD, fd, &_ev) != 0) {
```

Comment:
can mask be zero here?  We are triggered from something.. maybe an assertion that mask != 0?